### PR TITLE
calcbestshot fix - produces weird angles when shot origin is out of bounds

### PR DESCRIPTION
**Pull request by jared-paul**
### Description
Current calcbestshot implementation will produce warnings of "top angle < bottom angle" with the use of AngleSegment when the ball goes out of bounds. We depend on the shot origin being in the field bounds.

### Testing Done

Passes all tests with no warnings. Added an out of bounds test

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: src/software/ai/evaluation/calc_best_shot.cpp [hunk 1]
```diff
@@ -95,6 +95,14 @@ std::optional<Shot> calcBestShotOnGoal(const Field &field, const Team &friendly_
                                        const std::vector<Robot> &robots_to_ignore,
                                        double radius)
 {
+    if (shot_origin.x() < field.friendlyGoalCenter().x() ||
+        shot_origin.x() > field.enemyGoalCenter().x() ||
+        shot_origin.y() > field.enemyCornerPos().y() ||
+        shot_origin.y() < field.enemyCornerNeg().y())
```
Comment (jonathanlew) [inline src/software/ai/evaluation/calc_best_shot.cpp]: Can't we just use `contain(field.fieldLines(), shot_origin)`?

Comment (jared-paul) [reply src/software/ai/evaluation/calc_best_shot.cpp]: fieldLines() constructs a new Rectangle for every call which is pretty expensive relative to everything else. 

I was also thinking, should we only check x values? Is there a case (free kick for example) where the y value will be out of bounds?

Comment (jonathanlew) [reply src/software/ai/evaluation/calc_best_shot.cpp]: yeah x value should be fine. Please add a test for out of bounds x and out of bounds y and make sure we don't see the top angle bottom angle warning

**Review by jonathanlew — APPROVED** [review]

**Comment by HaomiaoZ** [comment]
I don't think out of bound was considered when I port over the code. It is great that you guys caught that. Have you tested with the case of out of bound in y direction?

**Comment by jonathanlew** [comment]
> I don't think out of bound was considered when I port over the code. It is great that you guys caught that. Have you tested with the case of out of bound in y direction?

No problem since the ball won't be going behind the goal line

**Comment by HaomiaoZ** [comment]
> > I don't think out of bound was considered when I port over the code. It is great that you guys caught that. Have you tested with the case of out of bound in y direction?
> 
> No problem since the ball won't be going behind the goal line

Wait I thought out of bound in y direction means the ball went out of bound from sidelines? Anyway as long as there is no warning when the ball is out of bound in y direction then it should be good. An extra y direction out of bound test might be a good idea.

**Review by akhilveeraghanta — APPROVED** [review]

**Pull request merged**